### ISSUE TITLE
Allow vertical scrolling beyond the last line

### ISSUE THREAD

**Issue by ygoe** [notepad-plus-plus/notepad-plus-plus#880]
Currently, you can only scroll the document so far that the last line of the document is displayed at the bottom of the window. Most applications give you a little more room to type and let you scroll so far that the last line of the document is displayed at the _top_ of the window. Notepad++ should do the same.

**Comment by dail8859** [comment]
There is a plug-in called ScrollPastEOF that does this. Would be nice to have a built in option for this though.

**Comment by aaaxx** [comment]
:+1:

**Comment by h-h-h-h** [comment]
@dg9ngf:
> Most applications give you a little more room to type and let you scroll so far that the last line of the document is displayed at the top of the window.

Most? I only know a few, and mostly I find it strange. But I see that it can make sense in a few scenarios. So an option to enable this would be okay to me.

> room to type

Do you mean something like going past the end of the line (column mode) without changing the view of the scrollbar? (E.g. with <kbd>Alt</kbd>+<kbd>Arrow key</kbd>, see issue https://github.com/notepad-plus-plus/notepad-plus-plus/issues/724 .) If not--not seeing the exact end of the file using the normal text navigation and the line numbers is negative from my point of view.

> let you scroll so far that the last line of the document is displayed at the top of the window.

This extent is what I find strange. Some programs allow to show the end of the document at the vertical center of the textbox. How about that?

**Comment by dg9ngf** [comment]
Well, I mostly know applications that do scroll beyond the last line. Visual Studio, TextPad, Word (in Draft view)... Windows Notepad and most GUI text boxes (which Notepad is likely based on) on the other hand do not. I don't use Notepad though so it has no influence on me.

Room to type means that I don't stick in the last line of the view with every line I type. I don't like typing in such tight space and prefer a little more room below. I'm not talking about "virtual lines" or "virtual cursor" which could go further to the right than the length of the current line. I'm not using that. Last time I've seen that was in DOS editors or IDEs (was it Turbo Pascal?).

Centre? Why that? I have absolutely really never seen such behaviour. It's either the top or the bottom. Anything else wouldn't make any sense IMO.

I see one point however that you could argue about: The scroll bar. When not scrolling beyond the last line, the scroll bar properly represents the currently visible part of the document relative to the entire document. That's how scoll bars work since Windows 95 and it's really practical. TextPad for example even keeps it that way when scrolling beyond the last line. The scroll bar is still as long as the viewport but sticks at the bottom and doesn't move any further down while scrolling the last line further up. Scintilla alters the scroll bar as soon as that extended scrolling is enabled: It effectively extends the document size by one viewport height and keeps moving the scroll bar further down as the last line is scrolling up the screen. This has the probably undesirable effect that the scroll bar does not touch the bottom when the last line of the document is displayed at the bottom of the screen. This gives you the impression that the document goes on when it really doesn't. Fixing this behaviour might need a change in Scintilla itself.

BTW, the plugin works fine so I'm using that for now. Based on the large number of plugins for any situation, I believe the use of plugins is common to NPP so this might already be the solution. I'm not sure whether an integrated option would be preferable. At least, the plugin description reads as if it's just a simple Scintilla switch, so it should be easy to implement. It would probably be easier to find as an integrated option.

**Comment by h-h-h-h** [comment]
**I think the behavior should be:**
- The scrollbar shows the real plus the artificial document range.
- You have the option to set the number of lines to add for scrolling (alternatively the percentage of the viewport height to add, rounded to number of lines). This is to not be annoyed by a too far scrolling; mostly you still want to see your document when scrolling to the virtual end (see also below).
- A slightly different background color where no document is, or, alternatively, a horizontal line (like column indicator) would be desirable.
- When pressing <kbd>Ctrl</kbd>+<kbd>End</kbd>, you get to the end of the line just as it is now. When you press it again, the scrolling shows all additional lines (like <kbd>End</kbd>, <kbd>End</kbd> on wrapped lines). The cursor position stays the same.
- When you type to the bottom of the textbox and press <kbd>Enter</kbd>, all additional lines are shown.
- If scrolling with the above two points would result in the second to last line being scrolled out of view, the scrolling is adjusted to show the second to last line on top and the last line with the cursor below it.

**Comment by dg9ngf** [comment]
* I'm fine with the added visual indication although I wouldn't request it.
* Ctrl+End goes to the end of the document, not the end of the line. Also, I wouldn't change that hotkey in any way. Explicit scrolling is done with Ctrl+Up/Down and this works as expected with the plugin installed.
* No scroll jumps when typing, please. You're adding too much unexpected logic that might not even be requested by anyone. I'd just keep it simple and limit the work to what's needed now. Again, other editors don't feature any of the additional things you listed.

**Comment by h-h-h-h** [comment]
> Ctrl+End goes to the end of the document, not the end of the line.

That was a mistake. [I corrected it.](https://github.com/notepad-plus-plus/notepad-plus-plus/issues/880#user-content-correction_1)

> I wouldn't change that hotkey in any way.

Deriving the described behavior from pressing <kbd>End</kbd> twice on a wrapped line is a great solution in my opinion.

> No scroll jumps when typing, please. You're adding too much unexpected logic that might not even be requested by anyone.

Don't you want <q>room to type</q> again after you used all <q>room to type</q> previously available? Otherwise we would have the same behavior as now, again&mdash;once you reached the bottom you constantly stay there, letting <q>scroll jumps</q> of a height of one line happen.

**Comment by dg9ngf** [comment]
Okay, if I don't press Ctrl+End twice, I won't be affected by a change in that hotkey. I don't mind. Maybe I'll even find it useful one day... ;-)

I want to be able to get that room to type. If the editor constrains the scrolling to the document length, it is impossible to get that room. If I can scroll further, it is possible an I'm close to happy. I only need to actually scroll down to get what I want. All editors I know do it like that. When the editor starts being smart and jump me around, I might lose orientation because I may not want to go so far as to see nothing above the current line anymore. So just don't implement that thinking for me, it saves you the work. :-)

Basically, all I'm asking for is the integration of the ScrollPastEOF plugin, and adjust the scrollbar so that it behaves like this plugin wasn't installed. When scrolled down to what is possible without the plugin now, the scrollbar is at the bottom end. When scrolling further, it should just not change anymore.

**Comment by h-h-h-h** [comment]
> When the editor starts being smart and jump me around, I might lose orientation because I may not want to go so far as to see nothing above the current line anymore. So just don't implement that thinking for me, it saves you the work.

The described <q>jumping behavior</q> is only addressing edge cases (e.g., when having a small window height). Imagine having Notepad++ maximized, 40 lines visible and the option of additional scrolling lines set to 10. You type to the bottom of the textbox, hit <kbd>Enter</kbd>, and get a room of 10 lines again, still seeing 30 lines of the ones previously typed. You could still turn that option off (or not on).

Do you see that a percentage option might be better than a fixed number of additional scrolling lines? I'm unsure about that.

> When scrolled down to what is possible without the plugin now, the scrollbar is at the bottom end. When scrolling further, it should just not change anymore.

But then you loose the ability to scroll this far by means of dragging the scrollbar rectangle. This way your only options to scroll this far would be to use <kbd>Ctrl</kbd>+<kbd>Up</kbd>/<kbd>Down</kbd> or the down button of the scrollbar. This is inconsistent behavior for me.

**Comment by dg9ngf** [comment]
As I said, I wouldn't want to set the additional space to anything other than 100% of the viewport height. Or to be very precise, I want the last line of the document to be in the first visible line. Not above or below that (we don't have pixel scrolling, so it's easy to define that). Smaller values just feel arbitrary and unnatural to me.

> But then you loose the ability to scroll this far by means of dragging the scrollbar rectangle.

Correct and accepted. Plus you have the mouse wheel. That's still plenty of options. That's how TextPad works for example. Visual Studio, on the other hand, does exactly what the ScrollPastEOF plugin does now. You just don't get a sense of the length of the document when the scrollbar doesn't indicate it correctly.

> This is inconsistent behavior for me.

So is extending the scroll space by a fixed number of lines for me. BTW, did you consider to limit that configured number to the actual height of the viewport? I really don't want to lose the end of the document way outside the viewport in small windows...

**Comment by h-h-h-h** [comment]
> As I said, I wouldn't want to set the additional space to anything other than 100% of the viewport height. Or to be very precise, I want the last line of the document to be in the first visible line. Not above or below that (we don't have pixel scrolling, so it's easy to define that). Smaller values just feel arbitrary and unnatural to me.

A percentage option&mdash;<q>How much of the textbox height should be added for additional scrolling?</q>&mdash;would make both of us happy. You would set it so 100 %, I would set it to 10 %. Smaller values are relevant for pressing <kbd>Ctrl</kbd>+<kbd>End</kbd> twice and the like.

> You just don't get a sense of the length of the document when the scrollbar doesn't indicate it correctly.

That's a reasonable point. I was just referring to software that implements it the way I described it.

> did you consider to limit that configured number to the actual height of the viewport? I really don't want to lose the end of the document way outside the viewport in small windows...

Sure, I addressed it with these words:

> @h-h-h-h:
If scrolling with the above two points would result in the second to last line being scrolled out of view, the scrolling is adjusted to show the second to last line on top and the last line with the cursor below it.

This way, the end of the document would never be scrolled out of view.

**Comment by dg9ngf** [comment]
Hm, I'd not consider the contents of a line. If the last line is empty, that could still be scrolled to the top. If you create a document of a thousand empty lines today, you could scroll all the way through it and would never see any non-empty content. It's nothing else.

The rest sounds good to me. Just out of curiosity, are you an active developer of NPP or more of the "community manager" type?

**Comment by h-h-h-h** [comment]
> I'd not consider the contents of a line. If the last line is empty, that could still be scrolled to the top.

Okay. Let's see what others say.

> Just out of curiosity, are you an active developer of NPP or more of the "community manager" type?

I'm just a user with a number of suggestions, hoping they will be accepted. If there are other suggestions I like, I review them. If you are willing to review my most important issues, here are some links: https://github.com/notepad-plus-plus/notepad-plus-plus/issues/452, https://github.com/notepad-plus-plus/notepad-plus-plus/issues/563, https://github.com/notepad-plus-plus/notepad-plus-plus/issues/778, https://github.com/notepad-plus-plus/notepad-plus-plus/issues/776, https://github.com/notepad-plus-plus/notepad-plus-plus/issues/572.

**Comment by dail8859** [comment]
Closed by 9fe39017cb7dd10cb6be82184129b9756d58e233
